### PR TITLE
refactor(date-input): use hooks

### PR DESCRIPTION
**Pull request by montezume**
#### Summary

Refactors the `DateInput` component to use hooks.

### CODE REVIEW

File: src/components/inputs/date-input/date-input.js [hunk 1]
```diff
@@ -25,267 +25,237 @@ import {
...
-  };
+    [props.id, props.name]
+  );
+
+  const handleChange = React.useCallback(
+    date => {
+      inputRef.current.setSelectionRange(0, 100);
+      emit(date);
+    },
+    [inputRef]
```
Comment (montezume) [inline src/components/inputs/date-input/date-input.js]: I'm not sure if this makes sense 🤔

File: src/components/inputs/date-input/date-input.js [hunk 2]
```diff
@@ -25,267 +25,237 @@ import {
...
-  inputRef = React.createRef();
-
-  state = {
-    calendarDate: this.props.value || getToday(),
-    suggestedItems: [],
-    highlightedIndex:
-      this.props.value === '' ? null : getDateInMonth(this.props.value) - 1,
-  };
-
-  showPrevMonth = () => {
```
Comment (montezume) [inline src/components/inputs/date-input/date-input.js]: this was never used

File: src/components/inputs/date-input/date-input.js [hunk 3]
```diff
@@ -25,267 +25,237 @@ import {
...
+              if (date === '') {
+                setSuggestedItems([]);
+                setHighlightedIndex(null);
               } else {
-                // input changed because user selected a date
-                this.setState({
-                  suggestedItems: [],
-                  highlightedIndex: null,
-                });
+                setSuggestedItems([date]);
```
Comment (montezume) [inline src/components/inputs/date-input/date-input.js]: I'm wondering if it makes sense to "batch" these updates? I guess one choice is to go with a "useReducer" but I feel like the readability is a bit lost then.

File: src/components/inputs/date-input/date-input.js (hunk 1 above)
Comment (tdeekens) [reply src/components/inputs/date-input/date-input.js]: To my understanding yes. useCallback does not have smart exceptions like any value of props or any ref. It is just a closure and memoization as far as I understood it.

File: src/components/inputs/date-input/date-input.js (hunk 3 above)
Comment (tdeekens) [reply src/components/inputs/date-input/date-input.js]: Won‘t React batch them? Not 100% sure but maybe.

File: src/components/inputs/date-input/date-input.js [hunk 4]
```diff
@@ -25,267 +25,233 @@ import {
...
-        // when cleared the value is null, but we always want it to be an
-        // empty string when there is no value.
-        value: value || '',
-      },
-    });
+  const jumpMonth = amount => {
+    const nextDate = changeMonth(calendarDate, amount);
+    setCalendarDate(nextDate);
+    setHighlightedIndex(0);
+  };
```
Comment (tdeekens) [inline src/components/inputs/date-input/date-input.js]: Is there a specific reason to not useCallback these two?

File: src/components/inputs/date-input/date-input.js [hunk 5]
```diff
@@ -25,267 +25,235 @@ import {
...
-  };
-
-  showToday = () => {
-    const today = getToday();
-    this.setState(
-      prevState => ({
-        calendarDate: today,
-        highlightedIndex:
-          prevState.suggestedItems.length + getDateInMonth(today) - 1,
+  const { onChange } = props;
```
Comment (emmenko) [inline src/components/inputs/date-input/date-input.js]: Same here

File: src/components/inputs/date-input/date-input.js [hunk 6]
```diff
@@ -25,267 +25,235 @@ import {
...
+      onBlur({
         target: {
-          id: this.props.id,
-          name: this.props.name,
+          id: props.id,
+          name: props.name,
         },
       });
-  };
+  }, [onBlur, props.id, props.name]);
```
Comment (emmenko) [inline src/components/inputs/date-input/date-input.js]: And here

File: src/components/inputs/date-input/date-input.js (hunk 5 above)
Comment (tdeekens) [reply src/components/inputs/date-input/date-input.js]: Yeah, this is how I would go about it too. Thanks.

**Pull request merged**